docs/developer-tools/Guard.md

```diff
@@ -0,0 +1,110 @@
+{
+    if (array is null)
+    {
+        throw new ArgumentNullException(nameof(array), "The array must not be null");
+    }
+
+    if (array.Length >= 10)
+    {
+        throw new ArgumentException($"The array must have less than 10 items, had a size of {array.Length}", nameof(array));
+    }
+
+    if (index < 0 || index >= array.Length)
+    {
+        throw new ArgumentOutOfRangeException(nameof(index), $"The index must be in the [0, {array.Length}) range, was {index}");
+    }
+
+    if (span.Length < array.Length)
+    {
+        throw new ArgumentException($"The target span is shorter than the input array, had a length of {span.Length}", nameof(span));
+    }
+
+    if (string.IsNullOrEmpty(text))
+    {
+        throw new ArgumentException("The input text can't be null or empty", nameof(text));
+    }
+}
+```
+
+And here is the same method, but using the new `Guard.APIs` to validate the input arguments:
+
+```csharp
+public static void SampleMethod(int[] array, int index, Span<int> span, string text)
+{
+    Guard.IsNotNull(array, nameof(array));
+    Guard.HasSizeAtLeast(array, 10, nameof(array));
+    Guard.IsInRangeFor(index, array, nameof(index));
+    Guard.HasSizeLessThanOrEqualTo(array, span, nameof(span));
+    Guard.IsNotNullOrEmpty(text, nameof(text));
+}
+```
+
+The `Guard` APIs will perform the required checks in the fastest way possible, and will throw the appropriate exception with a well formated message if they fail.
+
+## Methods
+
+There are dozens of different APIs and overloads in the `Guard` class, here are a few of them:
+
+| Methods | Return Type | Description |
+| -- | -- | -- |
+| IsNotNull<T>(T?, string) | void | Asserts that the input value is not null |
+| HasSizeAtLeast<T>(T[], int, string) | void | Asserts that the input array must have a size of at least or equal to a specified value |
+| IsInRangeFor<T>(int, T[], string) | void | Asserts that the input index is valid for a given array |
+| HasSizeLessThanOrEqualTo<T>(T[], T[], string) | void | Asserts that the source array must have a size of less than or equal to that of the destination array |
+| IsNotNullOrEmpty(string, string) | void | Asserts that the input string instance must not be null or empty |
+| IsOfType<T>(object, string) | void | Asserts that the input value is of a specific type |
+| IsReferenceEqualTo<T>(T, T, string) | void | Asserts that the input value must be the same instance as the target value |
+| IsTrue(bool, string) | void | Asserts that the input value must be true |
+| IsEqualTo<T>(T, T, string) | void | Asserts that the input value must be equal to a specified value |
+| IsLessThan<T>(T, T, string) | void | Asserts that the input value must be less than a specified value |
+| IsLessThanOrEqualTo<T>(T, T, string) | void | Asserts that the input value must be less than or equal to a specified value |
+| IsInRange<T>(T, T, T, string) | void | Asserts that the input value must be in a given range |
```

Comment:
Could be helpful for these ones to have small examples and show exactly which the order of the arguments are for?